### ISSUE TITLE
[Question] How to extend the functionality of \Codeception\Test\Unit?

### ISSUE THREAD

**Issue by x2ADLcRoo33ZygUA** [Codeception/Codeception#5689]
**TL;DR:**
In all of my unit tests, I use the same method. Now I paste it into each test using the copy-paste method, which, of course, is a very bad solution. How can I add that by default in every tester class?

---

Hello. I am newbie in testing and am not sure that my approach is generally correct, so I’ll take into account any of your suggestions for optimizing the process, but by now I’m testing the code as follows.

I am using **codeception** + **specify** + **verify** for a cross between unit testing and BDD. I will demonstrate with an example:

1. My codebase contains the `Email` class. The `Email` class contains the `Email::genLetterRaw()` method, which generates raw email letter.
2. I create a unit test with the command `./vendor/codeception/codeception/codecept generate:test unit Email`
3. I add the following code to the created tester class:
```
public function testgenLetterRaw() {
	$this->specify("I forgot to add recipients and pending an exception.", function () {
		// ...code
	});

	$this->specify("I expect the default value for the MIME-Version header is 1.0.", function () {
		// ...code
	});

	// ...etc
}
```

That is, 1 method of the tester class checks 1 class method from the code base, but the tester method contains several checks created using Specify.

This leads to the fact that quite often many `$this->specify()` callbacks have very similar contents, only the values of some parameters change.

I know the existence of [specify examples](https://github.com/Codeception/Specify#examples-dataproviders-alternative), but I do not like them for the following reasons:

- Parameter values are passed in an ordered rather than an associative array. If there are many values, it is difficult to immediately remember which parameter is responsible for what.
- I can’t set different messages for each data set (by message I mean the first argument `$ this->specify()`, like "*I expect...*")

Therefore, I use my own wrapper method, which with the help of foreach passes through the associative array and inside the loop calls `$this->specify()`.

**My question is this:** how is it most correct to extend the class `\Codeception\Test\Unit` in the framework of codeception and add the method I use to it? Now I am doing a copy-paste of this method in every tester class. This, of course, is the stupidest decision.

**Comment by Fenikkusu** [comment]
Let's see if I can get this right.

There are a few different ways to do this, depending on your specific needs. The first/simplest way is to create a new trait in `_support` and add a use statement to the file. This is a more traditional way. Another similar option is to create an abstract class (also in `_support`) and have it extend codeception's unit class. Then you just update all your unit classes to extend yours instead of codeception's. 

The more 'official' way of doing this though is going to be using the Actor's. In the `UnitTester` class (also in `_support`), you can add the method. Then in your unit test, you can access the method using `$this->tester->myMethod()`. IE:

```php
class UnitTester {
    //...

   public function myMethod()
   {
      //Do Something
   }
}

//=======

class SomethingTest extends \Codeception\Test\Unit {
    /** @var UnitTester */
    protected $tester;

    public function testSomething()
    {
        $this->tester->myMethod();
    }
}
```

Now, if you really wanted to go all out, you could do a custom module. This gives you the most flexibility, but comes at the cost of greater configuration. Based off what you have said, you're probably better off using the Actor.

**Comment by x2ADLcRoo33ZygUA** [comment]
@Fenikkusu 
I apologize for such a sl-o-o-o-o-w reaction.

In general, you understood me correctly, but the variant using `$tester` does not suit me for the following reason. I work with `specify`, respectively, i call `$this->specify()`. If I do it from `$tester`. then `$this` will obviously be an object of the `UnitTeste` class, not a `testgenLetterRaw`.

It seems to me that using `trait` would be a good idea, but could you clarify how to do this? I created the file in the `_support` folder, but it didn’t lead to anything. I think problems with namespace. Or is the fact that it needs to be additionally connected?

Thank you for your attention.

**Comment by Fenikkusu** [comment]
You may still want to use the `$tester` version. I feel like the scope will be in the unit, not the tester.  It has been a while since I tried similar, so I could be wrong. I personally haven't used specify, as I felt if extra overhead that didn't make sense to me, so you could also just not use specify. 

In terms of the trait, where you put the class doesn't really matter. The main thing you may need to do is add a `autoload-dev` section to your composer.json  & run composer dump-autoload so that it can find the class. IE: `autoload-dev: { "classmap" : "tests/_support/Traits" }`

**Comment by x2ADLcRoo33ZygUA** [comment]
Ultimately, I solved the problem using traits.

1. Create file `tests/_support/AwesomeTrait.php`
2. Added the following code to it
	```
	namespace Codeception;

	trait Awesome {
		// code
	}
	```
3. Added a line to the unittest file `use \Codeception\Awesome;`
4. In `composer.json`
	```json
	"autoload-dev": {
	  "classmap": ["tests/_support/AwesomeTrait.php"]
	}
	```
5. Run `composer dump-autoload`
6. Use `$this->MyFunctionName()`

Thanks again for your help.
